### PR TITLE
[MODFEE-83] Create pay endpoint

### PR DESCRIPTION
**Pull request by OleksandrVidinieiev**
Resolves [MODFEE-83](https://issues.folio.org/browse/MODFEE-83).

### CODE REVIEW

File: ramls/accounts.raml [hunk 1]
```diff
@@ -108,3 +110,31 @@ resourceTypes:
             body:
               text/plain:
                 example: "Internal server error"
+    /pay:
+      post:
+        is: [validate]
```
Comment (roman-barannyk) [inline ramls/accounts.raml:115]: Do we need it? It seems like we use java validation only

File: ramls/actionRequest.json [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      "$ref": "uuid.json",
+      "description": "Unique ID of the service point where the payment was made"
+    },
+    "userName": {
+      "type": "string",
+      "description": "Name of the user that was logged in when the action was performed"
+    },
+    "paymentMethod": {
+      "type": "string",
+      "description": "Name of the payment method, e.g. `Cash` or `Credit card`"
```
Comment (roman-barannyk) [inline ramls/actionRequest.json:33]: Does that make sense to define enum here?

File: src/main/java/org/folio/rest/service/ActionService.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  }
+
+  private Future<ActionContext> sendPatronNotice(ActionContext context) {
+    if (isTrue(context.getRequest().getNotifyPatron())) {
+      patronNoticeService.sendPatronNotice(context.getFeeFineAction());
+    }
+    return succeededFuture(context);
+  }
+
+  public static class ActionContext {
```
Comment (roman-barannyk) [inline src/main/java/org/folio/rest/service/ActionService.java:124]: Do you use it from another class?

File: ramls/accounts.raml (hunk 1 above)
Comment (OleksandrVidinieiev) [reply ramls/accounts.raml:115]: We use java-validation just for one field. RMB should validate the rest.

File: ramls/actionRequest.json (hunk 2 above)
Comment (OleksandrVidinieiev) [reply ramls/actionRequest.json:33]: No, payment methods are custom (created by users).

File: src/main/java/org/folio/rest/service/ActionService.java (hunk 3 above)
Comment (OleksandrVidinieiev) [reply src/main/java/org/folio/rest/service/ActionService.java:124]: Yes, `AccountsAPI`.

Comment (roman-barannyk) [reply src/main/java/org/folio/rest/service/ActionService.java:124]: I can't find using ActionContext in AccountsAPI

Comment (OleksandrVidinieiev) [reply src/main/java/org/folio/rest/service/ActionService.java:124]: https://github.com/folio-org/mod-feesfines/blob/9c491a976bac5636c85b55fdef3e655c471b448d/src/main/java/org/folio/rest/impl/AccountsAPI.java#L403

File: descriptors/ModuleDescriptor-template.json [hunk 4]
```diff
@@ -218,6 +218,35 @@
...
+            "inventory-storage.locations.item.get",
+            "inventory-storage.locations.collection.get",
+            "inventory-storage.location-units.libraries.item.get",
+            "inventory-storage.location-units.libraries.collection.get",
+            "inventory-storage.location-units.campuses.item.get",
+            "inventory-storage.location-units.campuses.collection.get",
+            "inventory-storage.location-units.institutions.item.get",
+            "inventory-storage.location-units.institutions.collection.get",
+            "users.item.get",
+            "users.collection.get"
```
Comment (alexanderkurash) [inline descriptors/ModuleDescriptor-template.json:247]: New permission set?

File: ramls/accounts.raml [hunk 5]
```diff
@@ -43,10 +45,10 @@ resourceTypes:
       pageable,
       facets
     ]
-    description: Return a list of accounts
+    description: "Return a list of accounts"
   post:
     is: [validate]
-    description: Create a account
+    description: "Create a account"
```
Comment (alexanderkurash) [inline ramls/accounts.raml:51]: an

File: ramls/actionRequest.json [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  "description": "Action request",
+  "type": "object",
+  "properties": {
+    "amount": {
+      "type": "string",
+      "description": "Action amount"
+    },
+    "comments": {
+      "type": "string",
+      "description": "Additional info for staff or patron. Format: [STAFF : staff comment \n PATRON : patron comment]"
```
Comment (alexanderkurash) [inline ramls/actionRequest.json:12]: How does this look in the docs?

File: ramls/actionResponse.json [hunk 7]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (alexanderkurash) [inline ramls/actionResponse.json:1]: Can we split this in two separate responses - one for success and one for failure?

File: src/main/java/org/folio/rest/domain/Action.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public enum Action {
+  PAY("Paid partially", "Paid fully"),
+  WAIVE("Waived partially", "Waived fully"),
+  TRANSFER("Transferred partially", "Transferred fully"),
+  REFUND("Refunded partially", "Refunded fully"),
+  CANCELLED(null, "Cancelled as error");
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/domain/Action.java:13]: Reuse the existing status literals?

File: src/main/java/org/folio/rest/domain/Action.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+  private final String partialResult;
+  private final String fullResult;
+
+  private static final List<String> terminalStatuses = Arrays.stream(values())
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/domain/Action.java:18]: Note: full refund shouldn't cause account closure.

File: src/main/java/org/folio/rest/repository/ActionRepository.java [hunk 10]
```diff
@@ -45,4 +45,10 @@ public ActionRepository(PostgresClient pgClient) {
         .orElse(null)
       );
   }
+
+  public Future<String> save(Feefineaction feefineaction) {
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/repository/ActionRepository.java:49]: Rename this repo?

File: src/main/java/org/folio/rest/service/ActionService.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+  }
+
+  public Future<ActionContext> pay(String accountId, ActionRequest request) {
+    return performAction(Action.PAY, accountId, request);
+  }
+
+  private Future<ActionContext> performAction(Action action, String accountId,
+    ActionRequest request) {
+
+    return succeededFuture(new ActionContext(accountId, request, action))
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/service/ActionService.java:51]: Parameter ordering

File: src/main/java/org/folio/rest/service/ActionService.java [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+    return validationService.validate(context.getAccount(), amount)
+      .map(result -> context.withRequestedAmount(parseDouble(amount)));
+  }
+
+  private Future<ActionContext> createFeeFineAction(ActionContext context) {
+    ActionRequest request = context.getRequest();
+    Account account = context.getAccount();
+    Action action = context.getAction();
+    Double requestedAmount = context.getRequestedAmount();
+    double remainingAmount = account.getRemaining() - requestedAmount;
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/service/ActionService.java:76]: Null check

File: src/main/java/org/folio/rest/service/ActionService.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+
+    String actionType = remainingAmount == 0
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/service/ActionService.java:78]: May not be exactly 0 with Double, let's round to 0.000

File: src/main/java/org/folio/rest/service/ActionService.java [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+    return actionRepository.save(feeFineAction)
+      .map(context.withFeeFineAction(feeFineAction));
+  }
+
+  private Future<ActionContext> updateAccount(ActionContext context) {
+    final Feefineaction feeFineAction = context.getFeeFineAction();
+    final Account account = context.getAccount();
+
+    if (isTerminalStatus(feeFineAction.getTypeAction())) {
+      account.getStatus().setName(FeeFineStatus.CLOSED.getValue());
```
Comment (alexanderkurash) [inline src/main/java/org/folio/rest/service/ActionService.java:107]: Refund

File: src/test/java/org/folio/rest/impl/AccountsActionsAPITests.java [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+    final Account account = createAccount(accountBalanceBefore);
+    postAccount(account);
+
+    String expectedPaymentStatus = terminalAction ? action.getFullResult() : action.getPartialResult();
+    String expectedAccountStatus = terminalAction ? "Closed" : "Open";
+    String requestedAmountString = String.valueOf(requestedAmount);
+
+    final ActionRequest request = createRequest(requestedAmountString);
+
+    String actionId = payClient.post(toJson(request))
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/impl/AccountsActionsAPITests.java:124]: Rename to feeFineActionId

File: src/test/java/org/folio/rest/utils/ResourceClients.java [hunk 16]
```diff
@@ -16,6 +16,14 @@ public static ResourceClient accountsCheckWaiveClient(String accountId) {
     return new ResourceClient("/accounts/" + accountId + "/check-waive");
   }
 
+  public static ResourceClient accountsActionClient(String accountId, String action) {
+    return new ResourceClient(String.format("/accounts/%s/%s", accountId, action));
+  }
+
+  public static ResourceClient actionsClient() {
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/utils/ResourceClients.java:23]: feeFineActions?

File: src/test/java/org/folio/rest/utils/ResourceClients.java [hunk 17]
```diff
@@ -16,6 +16,14 @@ public static ResourceClient accountsCheckWaiveClient(String accountId) {
     return new ResourceClient("/accounts/" + accountId + "/check-waive");
   }
 
+  public static ResourceClient accountsActionClient(String accountId, String action) {
+    return new ResourceClient(String.format("/accounts/%s/%s", accountId, action));
+  }
```
Comment (alexanderkurash) [inline src/test/java/org/folio/rest/utils/ResourceClients.java:21]: Let's make 4 methods

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**